code-oss: init at 1.89.1

inclyc opened the pull request:
## Description of changes

This PR adds a package named `vscode-oss` that build https://github.com/microsoft/vscode from source. (Currently we have vscode + vscodium but they are not built by nix). 

Not only does this provide vscode desktop, but also remote-ssh server and web server in separated outputs. That is, users of this package can use `vscode-oss.reh` for their remote-ssh extension host, instead of downloading a large unknown binary from microsoft.

Electron, nodejs, are reused from nixpkgs.

The package also allows user changing the product.json, to customize their own vscode variant.

I don't know how to use by-name technology as well as reusing vscode directory, sorry ;)

Initially when I wrote these expressions, vscode is on 1.89.1 and I tested it. Automatic version bump, ~1.90.0,~ I think I can file another subsequent PRs.

Fixes: #26698
Closes: #86122

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

inclyc committed:
vscode-oss: init at 1.89.1

inclyc committed:
fixup common

infinisil commented:
Check out https://github.com/NixOS/nixpkgs/tree/master/pkgs/by-name#recommendation-for-new-packages-with-multiple-versions :)

inclyc committed:
fixup all-packages.nix

inclyc committed:
fixup finalAttrs

inclyc commented on pkgs/top-level/all-packages.nix:
```diff
@@ -35246,19 +35246,27 @@ with pkgs;
+    vscodium-fhs = vscodium.fhs;
+    vscodium-fhsWithPackages = vscodium.fhsWithPackages;
+  })
+  vscode vschde-fhs vscode-fhsWithPackages vscode-with-extensions
```
```suggestion
  vscode vscode-fhs vscode-fhsWithPackages vscode-with-extensions
```

type, sorry.

inclyc committed:
fixup typo

SuperSandro2000 commented:
Did you explore how hard it would be to do this for vscodium?

SuperSandro2000 commented on pkgs/applications/editors/vscode/common.nix:
```diff
@@ -0,0 +1,78 @@
+  };
+
+  meta = {
+    description = ''
```
meta.description should fit on a single line without being to long to fit in on search.nixos.org

inclyc replied:
> meta.description should fit on a single line without being to long to fit in on search.nixos.org

These are copied from previous nixpkgs version. https://github.com/NixOS/nixpkgs/blob/1032df05d070c7102304f71930b39c505ea226d6/pkgs/applications/editors/vscode/vscode.nix#L86

Should I update this then?

SuperSandro2000 commented on pkgs/applications/editors/vscode/common.nix:
```diff
@@ -0,0 +1,78 @@
+      Linux and macOS
+    '';
+    longDescription = ''
+      Open source source code editor developed by Microsoft for Windows,
```
longDescription should not repeat description, as they are often displayed very close to each other.

inclyc replied:
Same as above, these words have already presented in nixpkgs before:

https://github.com/NixOS/nixpkgs/blob/1032df05d070c7102304f71930b39c505ea226d6/pkgs/applications/editors/vscode/vscode.nix#L91

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss.nix:
```diff
@@ -0,0 +1,331 @@
+}:
+
+let
+  inherit (nodePackages) node-gyp;
```
Task for another PR: we should package this package outside of nodePackages

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss.nix:
```diff
@@ -0,0 +1,331 @@
+  };
+
+  # replaces esbuild's download script with a binary from nixpkgs
+  patchEsbuild = path: version: ''
```
I am a bit surprised that this isn't in common but nothing to bikesheed over

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss.nix:
```diff
@@ -0,0 +1,331 @@
+      '';
+
+      outputHashMode = "recursive";
+      outputHashAlgo = "sha256";
```
This is redundant with outputHash

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss.nix:
```diff
@@ -0,0 +1,331 @@
+    "out"
+
+    # Remote extension host, for vscode remote development
+    "reh"
```
I would write those out, then the outputs are longer but it is easier to understand what they are. As a long term vscode user that fiddled a lot with remote extensions in nixos wsl, i have never seen this abbreviation before.

inclyc replied:
> I would write those out, then the outputs are longer but it is easier to understand what they are. As a long term vscode user that fiddled a lot with remote extensions in nixos wsl, i have never seen this abbreviation before.

Could you suggest a new name? maybe "remote-extension-host" ?

I don't have strong opinion about this but personally I want to keep the name because there are same abbreviation in VSCodium community, and VSCode gulp task. (Yes, microsoft just named it `reh`)

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss.nix:
```diff
@@ -0,0 +1,331 @@
+    xorg.libxkbfile
+  ] ++ lib.optionals stdenv.isDarwin [ darwin.apple_sdk.frameworks.Cocoa ];
+
+  patches = map (name: ./vscode-oss/patches/${name}) (
```
I think we might need to use filterSource here to not copy the entire tree again. There should be an issue about this.

inclyc commented:
> Did you explore how hard it would be to do this for vscodium?

VSCodium seems to be a set of patches (which aims to remove some MS branding stuff) combined with a LOT of scripts. 

Patches: https://github.com/VSCodium/vscodium/tree/master/patches

maybe for nixpkgs we can just apply their patches, but not reusing their scripts. Some of them basically look scary:

https://github.com/VSCodium/vscodium/blob/ffd7e631b05feef1b9142adb9e63303549cf4e6c/package_linux_bin.sh#L37

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss/product.nix:
```diff
@@ -0,0 +1,408 @@
+  licenseFileName = "LICENSE.txt";
+  reportIssueUrl = "https://github.com/microsoft/vscode/issues/new";
+  nodejsRepository = "https://nodejs.org";
+  urlProtocol = "code-oss";
```
Doesn't this break erb links from eg gitlab?

inclyc replied:
I don't know what's the meaning of this field actually, basically reading some comments from https://github.com/microsoft/vscode-docs/issues/5215. As per my understanding, it seems to be OK with `code-oss`, right?

Pandapip1 replied:
I can attest that this does break some extensions. Please override it to make it `vscode`.

SuperSandro2000 commented on pkgs/applications/editors/vscode/vscode-oss/product.nix:
```diff
@@ -0,0 +1,408 @@
+  extensionAllowedBadgeProvidersRegex = [
+    "^https:\\/\\/github\\.com\\/[^/]+\\/[^/]+\\/(actions\\/)?workflows\\/.*badge\\.svg"
+  ];
+  extensionEnabledApiProposals = {
```
This seems rather aggressive to put this into nix code. Can't we read that from the downloaded source file?

inclyc replied:
> Can't we read that from the downloaded source file?

The source file is rather empty, and these items listed mainly comes from VSCodium community.

https://github.com/microsoft/vscode/blob/main/product.json

inclyc committed:
vscode-oss: 1.89.1 -> 1.90.0

Bump nodejs and electron version according to https://code.visualstudio.com/updates/v1_90#_electron-29-update

inclyc committed:
Revert "vscode-oss: 1.89.1 -> 1.90.0"

This reverts commit c7388816273bee7885111ac825f5e058033a039e.

code-oss> [05:53:03] 'compile-extensions-build' errored after 2.71 s
code-oss> [05:53:03] Error: Command failed: yarn list --prod --json
code-oss>     at genericNodeError (node:internal/errors:984:15)
code-oss>     at wrappedFn (node:internal/errors:538:14)
code-oss>     at ChildProcess.exithandler (node:child_process:422:12)
code-oss>     at ChildProcess.emit (node:events:518:28)
code-oss>     at ChildProcess.emit (node:domain:551:15)
code-oss>     at maybeClose (node:internal/child_process:1105:16)
code-oss>     at Socket.<anonymous> (node:internal/child_process:457:11)
code-oss>     at Socket.emit (node:events:518:28)
code-oss>     at Socket.emit (node:domain:551:15)
code-oss>     at Pipe.<anonymous> (node:net:337:12)
code-oss> error Command failed with exit code 1.

inclyc commented:
1.90.0 cannot be built from linux, revert it for now.

superherointj commented:
@inclyc I suggest you to keep commits squashed.

inclyc committed:
vscode-oss: do not patch esbuild

linsui commented:
Does this work with both extensions installed from nix and the market at the same time?

inclyc commented:
> Does this work with both extensions installed from nix and the market at the same time?

Yes.

batonac commented:
@inclyc you're a hero, I'm so happy for this!

Now that [1.90.2](https://github.com/microsoft/vscode/releases/tag/1.90.2) is out, could we try the latest version again?

inclyc commented:
Hi @batonac,

I have a personal testing [code-oss](https://github.com/inclyc/flakes/blob/main/pkgs/code-oss/package.nix) package in my flake and I recently bumped it to version 1.90.2.

I'd like to test it for days before updating this PR.

batonac commented:
Hey @inclyc, apparently this stalled?

inclyc commented:
> Hey @inclyc, apparently this stalled?

Hi, I'm testing this package in [my own flake]( https://github.com/inclyc/flakes/blob/main/pkgs/code-oss/package.nix)

inclyc commented:
Because microsoft/vscode has migrated to npm again, currently it is impossible to package vscode from source. Reasons are:

- It uses npm, and the offline cache produced by npm itself cannot be reproduced.
- `prefetch-npm-deps` cannot deal with such complex project, (i.e. vscode), there are many "package-lock.json" needs to be prefetched, but optional dependencies must be ignored, `prefetch-npm-deps` seems to be confused by that.
